### PR TITLE
T-25 Create asset valuation calculator

### PR DESCRIPTION
**Pull request by goremukin**
A class was created that gets the latest current currency rate. And a class that calculates at the rate. And tests

### CODE REVIEW

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/AssetCalculatorTest.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿using Moq;
+using NUnit.Framework;
+using PortfolioKeeper.BusinessLogic.Helpers;
+using System.Diagnostics.CodeAnalysis;
+
+namespace PortfolioKeeper.Tests.BusinessLogic.Helpers
+{
+    [ExcludeFromCodeCoverage]
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/AssetCalculatorTest.cs]: Use TestFixture here. remove ExcludeFromCodeCoverage

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/AssetCalculatorTest.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            exchangeRateProvider = new Mock<IExchangeRateProvider>();
+            exchangeRateProvider.Setup(moq => moq.GetExchangeRate(It.IsAny<char>(), It.IsAny<bool>())).Returns(1.32M);
+            underTest = new AssetCalculator(exchangeRateProvider.Object);
+        }
+
+        [Test]
+        public void AssetCalculator_When_ExchangeRateProvider_Is_Null_Should_Throw_ArgumentNullException()
+        {
+            Assert.That(() => new AssetCalculator(null), Throws.ArgumentNullException);
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/AssetCalculatorTest.cs]: use FluentAssertions

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/AssetCalculatorTest.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        [Test]
+        public void Calculate_Should_Correctly_Calculate()
+        {
+            Assert.AreEqual(3.96M, underTest.Calculate((char)1, 3));
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/AssetCalculatorTest.cs]: same. also check various values. use testcases

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+using PortfolioKeeper.Data.Entities;
+using PortfolioKeeper.Data.Interfaces;
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace PortfolioKeeper.Tests.BusinessLogic.Helpers
+{
+    [ExcludeFromCodeCoverage]
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs]: TestFixture

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        private Mock<IUnitOfWork> uof;
+        private Mock<IRepository<ExchangeRate, int>> exchangeRateRepository;
+        private ExchangeRateProvider underTest;
+
+        [SetUp]
+        public void Init()
+        {
+            exchangeRateRepository = new Mock<IRepository<ExchangeRate, int>>();
+            exchangeRateRepository.Setup(moq =>
+            moq.Get(It.IsAny<Expression<Func<ExchangeRate, bool>>>(),
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs]: brrr...

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+                      } }).Where(predicate.Compile()));
+            uof = new Mock<IUnitOfWork>();
+            uof.Setup(moq => moq.GetRepository<ExchangeRate, int>()).Returns(exchangeRateRepository.Object);
+            underTest = new ExchangeRateProvider(uof.Object);
+        }
+
+        [Test]
+        public void ExchangeRateProvider_When_UnitOfWork_Is_Null_Should_Throw_ArgumentNullException()
+        {
+            Assert.That(() => new ExchangeRateProvider(null), Throws.ArgumentNullException);
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs]: Fluent

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        [Test]
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs]: use ExpectedException attribute

File: PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                          FromCurrencyId = 2,
+                          RateDate = DateTime.Now,
+                          BuyRate = 0.02M
+                      } }).Where(predicate.Compile()));
+            uof = new Mock<IUnitOfWork>();
+            uof.Setup(moq => moq.GetRepository<ExchangeRate, int>()).Returns(exchangeRateRepository.Object);
+            underTest = new ExchangeRateProvider(uof.Object);
+        }
+
+        [Test]
```
Comment (stas-litvinenka) [inline PortfolioKeeper.Tests/BusinessLogic/Helpers/ExchangeRateProviderTest.cs]: ExpectedException attribute

File: PortfolioKeeper/BusinessLogic/Helpers/AssetCalculator.cs [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+        private IExchangeRateProvider provider;
+
+        public AssetCalculator(IExchangeRateProvider provider)
+        {
+            this.provider = provider ?? throw new ArgumentNullException(nameof(provider));
+        }
+
+        public decimal Calculate(char currencyId, decimal amount)
+        {
+            return provider.GetExchangeRate(currencyId) * amount;
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/AssetCalculator.cs]: var exchangeRate = provider.GetExchangeRate(currencyId);

return exchangeRate *amount;

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public class ExchangeRateProvider : IExchangeRateProvider
+    {
+        private readonly IUnitOfWork unitOfWork;
+        private readonly IRepository<ExchangeRate, int> exchangeRateRepository;
+
+        public ExchangeRateProvider(IUnitOfWork unitOfWork)
+        {
+            this.unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));
+            exchangeRateRepository = unitOfWork.GetRepository<ExchangeRate, int>() ?? throw new ArgumentNullException(nameof(exchangeRateRepository));
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: ?? throw new ArgumentNullException(nameof(exchangeRateRepository)); to new line

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+        }
+
+        public decimal GetExchangeRate(char currencyId, bool buyRate = true)
+        {
+            var result = exchangeRateRepository
+                .Get(e => e.FromCurrencyId == currencyId)
+                ?.OrderByDescending(e => e.RateDate)
+                ?.FirstOrDefault();
+
+            if (result == null)
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: InvalidOperationException - try something else. not best type of exception for this.

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+            this.unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));
+            exchangeRateRepository = unitOfWork.GetRepository<ExchangeRate, int>() ?? throw new ArgumentNullException(nameof(exchangeRateRepository));
+        }
+
+        public decimal GetExchangeRate(char currencyId, bool buyRate = true)
+        {
+            var result = exchangeRateRepository
+                .Get(e => e.FromCurrencyId == currencyId)
+                ?.OrderByDescending(e => e.RateDate)
+                ?.FirstOrDefault();
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: there can be more than 1 exchangeRate(we have few banks). so you need to take best price also. for latest day

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        private readonly IRepository<ExchangeRate, int> exchangeRateRepository;
+
+        public ExchangeRateProvider(IUnitOfWork unitOfWork)
+        {
+            this.unitOfWork = unitOfWork
+                ?? throw new ArgumentNullException(nameof(unitOfWork));
+            exchangeRateRepository = unitOfWork.GetRepository<ExchangeRate, int>() ?? throw new ArgumentNullException(nameof(exchangeRateRepository));
+        }
+
+        public decimal GetBestExchangeRate(char currencyId, bool buyRate = true)
```
Comment (EdwOK) [inline PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: Split the logic method

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+{
+    public class ExchangeRateProvider : IExchangeRateProvider
+    {
+        private readonly IUnitOfWork unitOfWork;
+        private readonly IRepository<ExchangeRate, int> exchangeRateRepository;
+
+        public ExchangeRateProvider(IUnitOfWork unitOfWork)
+        {
+            this.unitOfWork = unitOfWork
+                ?? throw new ArgumentNullException(nameof(unitOfWork));
```
Comment (EdwOK) [inline PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: single line

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        public decimal GetBestExchangeRate(char currencyId, bool buyRate = true)
+        {
+            var lastRate = exchangeRateRepository
+                .Get(e => e.FromCurrencyId == currencyId,
+                    orderBy: q => q.OrderByDescending(e => e.RateDate))
+                ?.FirstOrDefault();
+
+            if (lastRate == null)
+            {
+                throw new ArgumentException();
```
Comment (stas-litvinenka) [inline PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: please use more appropriate exception here

File: PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        private readonly IUnitOfWork unitOfWork;
+        private readonly IRepository<ExchangeRate, int> exchangeRateRepository;
+
+        public ExchangeRateProvider(IUnitOfWork unitOfWork)
+        {
+            this.unitOfWork = unitOfWork ?? throw new ArgumentNullException(nameof(unitOfWork));
+            exchangeRateRepository = unitOfWork.GetRepository<ExchangeRate, int>() ?? throw new ArgumentNullException(nameof(exchangeRateRepository));
+        }
+
+        public decimal GetBestExchangeRate(char currencyId, bool buyRate = true)
```
[the remark this answers is not in the document]
Comment (stas-litvinenka) [reply PortfolioKeeper/BusinessLogic/Helpers/ExchangeRateProvider.cs]: it does not matter. will be changed soon. but good point

**Pull request merged**